### PR TITLE
openssl3: new recipe

### PR DESCRIPTION
**Pull request by davidkaroly**
I guess this one is long overdue... the recipe might be not the most beautiful and it has seen only limited testing so far but at least it's a start.

Successful build on x86_64, x86_gcc and x86 secondary arch.

There are a few failed test cases from "make test" - I'll try to add more details on that soon.

So far I marked the recipe as untested which more or less covers the current situation, and this way it should not interfere with existing openssl-1.1.1w recipe.

### CODE REVIEW

File: dev-libs/openssl/openssl3-3.0.12.recipe [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+REVISION="1"
+SOURCE_URI="https://www.openssl.org/source/openssl-$portVersion.tar.gz"
+CHECKSUM_SHA256="f93c9e8edde5e9166119de31755fc87b4aa34863662f67ddfcba14d0b6b69b61"
+SOURCE_DIR="openssl-$portVersion"
+PATCHES="openssl3-$portVersion.patchset"
+
+ARCHITECTURES="?all"
+SECONDARY_ARCHITECTURES="?x86"
+
+libVersion=3
```
Comment (korli) [inline dev-libs/openssl/openssl3-3.0.12.recipe:27]:
```suggestion
libVersion="3"
```

Comment (davidkaroly) [reply dev-libs/openssl/openssl3-3.0.12.recipe]: done

**Comment by davidkaroly** [comment]
### Test case execution results on x86_64

openssl-1.1.1w

```
Test Summary Report
-------------------
../test/recipes/01-test_symbol_presence.t        (Wstat: 256 Tests: 4 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
../test/recipes/02-test_errstr.t                 (Wstat: 17408 Tests: 70 Failed: 68)
  Failed tests:  1-68
  Non-zero exit status: 68
../test/recipes/80-test_cms.t                    (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  4
  Non-zero exit status: 1
../test/recipes/80-test_ssl_new.t                (Wstat: 256 Tests: 30 Failed: 1)
  Failed test:  22
  Non-zero exit status: 1
Files=159, Tests=2644, 812 wallclock secs ( 3.16 usr  0.53 sys + 309.52 cusr 87.24 csys = 400.46 CPU)
Result: FAIL
```

openssl3-3.0.14

```
Test Summary Report
-------------------
01-test_symbol_presence.t        (Wstat: 256 Tests: 4 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
04-test_err.t                    (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
79-test_http.t                   (Wstat: 256 Tests: 2 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
80-test_cmp_http.t               (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 6 tests but ran 0.
80-test_cms.t                    (Wstat: 256 Tests: 18 Failed: 1)
  Failed test:  5
  Non-zero exit status: 1
80-test_ssl_new.t                (Wstat: 256 Tests: 30 Failed: 1)
  Failed test:  22
  Non-zero exit status: 1
Files=250, Tests=3287, 1325 wallclock secs (11.80 usr  1.88 sys + 746.99 cusr 161.22 csys = 921.90 CPU)
Result: FAIL

```

**Comment by davidkaroly** [comment]
### Brief analysis of the test case results.

* There are a few test cases that never seemed to work: symbol_presence, errstr, cms, ssl_new
* The additional failed test cases on 3.0.14 are all new test cases, none of them were there on 1.1.1.
* The set of failed test cases on x86 secondary arch is identical to that of x86_64

* 2 new test cases failed on 3.0.14 on all architectures: http, cmp_http. These might be related to socket options and/or IPv6.
* In addition to that, pkeyutl test failed on 3.0.14 x86_gcc2. This test case does what it says i.e. it invokes `openssl pkeyutl -verify`

**Comment by waddlesplash** [comment]
Does this work on beta4 or does it need some of the new things added in nightlies?

**Comment by davidkaroly** [comment]
> Does this work on beta4 or does it need some of the new things added in nightlies?

Good question. So far I tested on nightlies. I'll give it a try on beta4 as well.

**Comment by davidkaroly** [comment]
Successful build on beta4 for x86_64, x86_gcc2 and x86 secondary arch. A bunch of failed test cases but also for openssl-1.1.1. I haven't analyzed the test results in detail yet.

**Comment by diversys** [comment]
In the meantime there has been an openssl 3.2.0 release

**Comment by davidkaroly** [comment]
> In the meantime there has been an openssl 3.2.0 release

It seems that the LTS line is still 3.0:

_Note: The latest stable version is the 3.2 series supported until 23rd November 2025. Also [available](https://www.openssl.org/support/contracts.html) is the 3.1 series supported until 14th March 2025, and **the 3.0 series which is a Long Term Support (LTS) version and is supported until 7th September 2026**. All older versions (including 1.1.1, 1.1.0, 1.0.2, 1.0.0 and 0.9.8) are now out of support and should not be used. Users of these older versions are encouraged to upgrade to 3.2 or 3.0 as soon as possible. Extended support for 1.1.1 and 1.0.2 to gain access to security fixes for those versions is available._

**Comment by davidkaroly** [comment]
some more patching:
- link against zlib if zlib is enabled
- skip test case test_print_error_format - this seems to be a deprecated feature anyway and Haiku error codes are not handled very well by OpenSSL
- don't use IPV6_V6ONLY as it doesn't seem to work on Haiku (but maybe we should skip all IPv6 test cases anywa)

Test results on x86_64 nightly:
- before patching: 6 failed test cases - symbol_presence, errstr, http, cmp_http, cms, ssl_new
- after patching: 1 failed test case - cmp_http

**Comment by pulkomandy** [comment]
> and Haiku error codes are not handled very well by OpenSSL

Can this be fixed? Do we need to use posix_error_mapper (to avoid problems with our error codes being negative), or is the problem somewhere else? Not handling errors well seems a bit worrying for OpenSSL.

**Comment by waddlesplash** [comment]
Since OpenSSL can return errors to other applications I don't think we should use the error mapper here.

**Comment by davidkaroly** [comment]
AFAICT it never worked properly, even on openssl-1.1.1:


> ~> openssl x509 -in blabla
> Can't open blabla for reading, No such file or directory
> 1555177167584:error:02001003:system library:fopen:**No Error (3)**:crypto/bio/bss_file.c:69:fopen('blabla','r')
> 1555177167584:error:2006D080:BIO routines:BIO_new_file:no such file:crypto/bio/bss_file.c:76:
> unable to load certificate
> ~>
>

**Comment by davidkaroly** [comment]
One more patch: listening sockets shall default to IPv4 as AF_UNSPEC doesn't really work.
This gives us a PASS on x86_64 and x86 secondary arch.
Still a few failed test cases on x86_gcc2: pkeyutl, verify, evp, ssl_new.

**Comment by waddlesplash** [comment]
What else do we need here?

**Comment by kenmays** [comment]
For Haiku R1B5, may want to bump to:  openssl-3.0.13 (2024-Jan-30)

**Comment by davidkaroly** [comment]
> What else do we need here?

probably not much else needed, let me get back to this over the weekend.

meanwhile one question: what should be the package name - openssl3 or openssl_3.0? perhaps openssl3.0?

**Comment by waddlesplash** [comment]
Probably `openssl3`. The future 3.x versions will be backwards compatible, I think?

We should probably also switch to the most recent LTS version before merging this.

**Comment by davidkaroly** [comment]
re-running some tests:

openssl3-3.0.12 and 3.0.13 gives us a PASS on hrev57499 x86_64
however, test case results are raher ugly on more recent hrev57564.

I'm tempted to say that it's a Haiku Kernel issue as it's also apparent in the commit history that there's been much churn in the network stack recently.

Updated patch for openssl3-3.0.13 coming soon.

(side note: I'm using GitHub from Falkon now, looks quite good, except for the YubiKey based authentication which gives us a segmentation fault. Haven't tried recent builds of Epiphany yet)

**Comment by waddlesplash** [comment]
>  however, test case results are raher ugly on more recent hrev57564.

What fails? Any obvious problems?

**Comment by waddlesplash** [comment]
The primary network changes in that range are the TCP refactors. I guess I would be surprised if they actually broke anything. Even one new failed test case should be enough to dig in to exactly what error or unexpected behavior is occurring, and we can work from there.

**Comment by kenmays** [comment]
@davidkaroly - I uploaded the unit testing on Haiku R1B4 x64 for comparison to the Haiku nightly builds.

**Comment by waddlesplash** [comment]
> Not so easy to point out exactly what fails - there are like a dozen of failed test cases and some of them even getting blocked.

Should be fixed now after changes in Haiku?

**Comment by davidkaroly** [comment]
> > Not so easy to point out exactly what fails - there are like a dozen of failed test cases and some of them even getting blocked.
> 
> Should be fixed now after changes in Haiku?

I'll take one more look at

**Comment by davidkaroly** [comment]
Looks good on x86_64 hrev57666:

```All tests successful.
Files=253, Tests=3310, 636 wallclock secs ( 7.34 usr  1.82 sys + 451.52 cusr 135.70 csys = 596.38 CPU)
Result: PASS
make[2]: Leaving directory '/sources/openssl-3.0.13'
make[1]: Leaving directory '/sources/openssl-3.0.13'
cleaning 'chroot/boot' folder
keeping chroot folder /boot/home/haikuports/dev-libs/openssl/work-3.0.13 intact for inspection
~/haikuports/dev-libs/openssl> uname -a
Haiku shredder 1 hrev57666 Mar 24 2024 06:02:46 x86_64 x86_64 Haiku
~/haikuports/dev-libs/openssl>
```

**Pull request merged**